routing/base/bfs.hpp

```diff
@@ -0,0 +1,103 @@
+#pragma once
+
+#include <algorithm>
+#include <functional>
+#include <map>
+#include <queue>
+#include <set>
+#include <vector>
+
+namespace routing
+{
+template <typename Graph>
+class BFS
+{
+public:
+  using Vertex = typename Graph::Vertex;
+  using Edge = typename Graph::Edge;
+  using Weight = typename Graph::Weight;
+
+  struct State
+  {
+    State(Vertex const & v, Vertex const & p) : m_vertex(v), m_parent(p) {}
+
+    Vertex m_vertex;
+    Vertex m_parent;
+  };
+
+  explicit BFS(Graph & graph): m_graph(graph) {}
+
+  void Run(Vertex const & start, bool isOutgoing,
+           std::function<bool(State const &)> && onVisitCallback);
+
+  std::vector<Vertex> ReconstructPath(Vertex from, bool reverse);
+
+private:
+  Graph & m_graph;
+  std::map<Vertex, Vertex> m_parents;
+};
+
+template <typename Graph>
+void BFS<Graph>::Run(Vertex const & start, bool isOutgoing,
+                     std::function<bool(State const &)> && onVisitCallback)
+{
+  m_parents.clear();
+  std::set<Vertex> used;
+
+  std::queue<Vertex> queue;
+  queue.emplace(start);
+
+  std::vector<Edge> edges;
+  while (!queue.empty())
+  {
+    Vertex const current = queue.front();
+    queue.pop();
+
+    if (used.count(current) != 0)
+      continue;
+
+    used.emplace(current);
+
+    if (isOutgoing)
+      m_graph.GetOutgoingEdgesList(current, edges);
+    else
+      m_graph.GetIngoingEdgesList(current, edges);
+
+    for (auto const & edge : edges)
+    {
+      Vertex const & child = edge.GetTarget();
+      if (used.count(child) != 0)
```

Comment:
You're checking used both when adding a vertex and when removing it from a queue now. This seems excessive.
Besides, why not just reuse the ```parent``` map? There's a lot of ways to deal with the starting vertex (check explicitly, add it as a parent to itself and later remove, etc.)

Reply:
fixed